src/ng/compile.js

```diff
@@ -3508,21 +3508,20 @@ function $CompileProvider($provide, $$sanitizeUriProvider) {
             if (optional && !attrs[attrName]) break;
 
             parentGet = $parse(attrs[attrName]);
-            var deepWatch = parentGet.literal;
 
             var initialValue = destination[scopeName] = parentGet(scope);
             initialChanges[scopeName] = new SimpleChange(_UNINITIALIZED_VALUE, destination[scopeName]);
 
             removeWatch = scope.$watch(parentGet, function parentValueWatchAction(newValue, oldValue) {
               if (oldValue === newValue) {
-                if (oldValue === initialValue || (deepWatch && equals(oldValue, initialValue))) {
+                if (equals(oldValue, initialValue)) {
```

Comment:
I don't have any better idea 😞

The thing is we need to call `$onChanges` synchronously (while the watchAction will only fire asynchronously), so we call `$onChanges` manually. When the watchAction does fire for the first time we need to determine if the value has changed - i.e. if `initialValue` is equal to `oldValue` in the `$watch` sense. That's why we need to differentiate between literals (compared using `equals`) and non-literals (compared using `===`).